src/g_kbdnav.c

```diff
@@ -0,0 +1,2183 @@
+        t_object *obj = pd_checkobject(&gobj->g_pd);
+            /* get the object text */
+            char *buf;
+            int bufsize;
+            binbuf_gettext(obj->te_binbuf, &buf, &bufsize);
+            buf = t_resizebytes(buf, bufsize, bufsize+1);
+            buf[bufsize] = 0;
+        int type = mag->mt_iotype;
+        post("   magtrav[%d] OBJ NAME = [%s]", count, buf);
+        post("   magtrav[%d].mt_dist = %f", count, mag->mt_dist);
+        post("   magtrav[%d].mt_iotype = %d", count, mag->mt_iotype);
+        post("   magtrav[%d].type = %s", count, type == IO_OUTLET ? "outlet" : "inlet");
+        post("   magtrav[%d].mt_objindex = %d", count, mag->mt_objindex);
+        post("   magtrav[%d].mt_ioindex = %d", count, mag->mt_ioindex);
+        post("   magtrav[%d].mt_next = %lx", count, mag->mt_next);
+        ++count;
+    }
+    post("-----");
+}
+
+void initialize_kbdnav(t_kbdnav *x)
+{
+    x->kn_ioindex = 0;
+    x->kn_state = KN_INACTIVE;
+    x->kn_moddown = 0;
+    x->kn_connindex = 0;
+    x->kn_outconnect = NULL;
+    x->kn_highlight = "blue";
+    x->kn_linetraverser = (t_linetraverser *)getbytes(sizeof(*x->kn_linetraverser));
+    x->kn_magtrav = NULL;
+    x->kn_chosennumber = -1;
+}
+
+
+/* callled when the user enters keyboard navigation
+   by selecting an object and pressing mod+Up/down.
+   We activate the selected object's first inlet or outlet,
+   depending on the pressed arrow.
+   "io_type" describes if the user will start
+   navigation on an inlet or outlet ("IO_INLET" or "IO_OUTLET") */
+void kbdnav_activate_kbdnav(t_canvas *x, int io_type)
+{
+    t_kbdnav *kbdnav = canvas_get_kbdnav(x);
+    int n_of_selected_objs = kbdnav_count_selected_objects(x);
+    if( n_of_selected_objs == 1 )
+    {
+        kbdnav_activate_object( x, kbdnav_get_selected_obj(x), io_type );
+    }
+    else
+    {
+        kbdnav_deactivate(x);
+        return;
+    }
+}
+
+/* called when the user exits keyboard navigation */
+void kbdnav_deactivate(t_canvas *x)
+{
+    t_kbdnav *kbdnav = canvas_get_kbdnav(x);
+    kbdnav->kn_ioindex = 0;
+    kbdnav->kn_state = KN_INACTIVE;
+    kbdnav->kn_connindex = 0;
+    kbdnav->kn_indexvis = 0;
+    kbdnav->kn_selobj = 0;
+    kbdnav->kn_outconnect = NULL;
+    kbdnav->kn_chosennumber = -1;
+    kbdnav_magnetic_connect_free(x);
+    canvas_redraw(x);
+}
+
+
+/* parse key press/release events related to keyboard navigation*/
+int kbdnav_key(t_canvas *x, t_symbol *s, int ac, t_atom *av, int keynum, int down, int shift, t_symbol *gotkeysym)
+{
+    t_kbdnav *kbdnav = canvas_get_kbdnav(x);
+    if( !kbdnav ) return 1;
+
+    /* set our modifier state */
+#if defined(__APPLE__)
+    if( !strcmp(gotkeysym->s_name, "Meta_L") )
+#else
+    /* TCL reports Right Control keyup events as Control_L */
+    /* https://wiki.tcl-lang.org/page/Modifier+Keys*/
+    if( ( !strcmp(gotkeysym->s_name, "Control_L") || !strcmp(gotkeysym->s_name, "Control_R")) )
+#endif
```

Comment:
i think it would be better to do the actual keybindings on the Tcl/Tk side.
e.g. rather than checking whether the pd-core received a string like "Control_L", "Control_R" or "Meta_L" (btw, what happened to "Meta_R"?) you should instead create bindings for for `<$::modifier-Key-Up>` and the like.
this will also automatically handle any platform differences in the "Meta"-key

once the bindings are handled in Tcl/Tk, add  one (or sevaral) methods to the canvas-class that explicitly deal with keyboard-patching.
you could also give the GUI a notion of whether the system is currently in keyboard-patching mode, in which case ordinary keystrokes that cannot properly be bound (like <kbd>g</kbd>) might get re-directed.

Reply:
Yes.

While rewriting the kbd nav code to use PD's class/method paradigm maybe all the kbdnav bindings could be handler gui side and call dedicated methods.

Maybe something like

**Gui:**
```TCL
proc ::pd_bindings::kbdnav_bindings {}{
   [...]
   bind all <$::modifier-Key-Left> { kbdnav_left %W }
   [...]
}
```
**Core:**

```C
void g_canvas_setup(void)
{
   // [...]
   /* -------------------------- keyboard navigation ----------------------------- */
   class_addmethod(canvas_class, (t_method)kbdnav_leftkey, gensym("kbdnav_left"),
                       A_GIMME, A_NULL);
   // [...]
}
```
```